sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala

```diff
@@ -90,13 +92,42 @@ object SQLExecution {
    * thread from the original one, this method can be used to connect the Spark jobs in this action
    * with the known executionId, e.g., `BroadcastExchangeExec.relationFuture`.
    */
-  def withExecutionId[T](sc: SparkContext, executionId: String)(body: => T): T = {
+  def withExecutionId[T](sparkSession: SparkSession, executionId: String)(body: => T): T = {
+    val sc = sparkSession.sparkContext
     val oldExecutionId = sc.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    withSQLConfPropagated(sparkSession) {
+      try {
+        sc.setLocalProperty(SQLExecution.EXECUTION_ID_KEY, executionId)
+        body
+      } finally {
+        sc.setLocalProperty(SQLExecution.EXECUTION_ID_KEY, oldExecutionId)
+      }
+    }
+  }
+
+  def withSQLConfPropagated[T](sparkSession: SparkSession)(body: => T): T = {
+    // Set all the specified SQL configs to local properties, so that they can be available at
+    // the executor side.
```

Comment:
properties are serialized per task.  how unusual would it be for there to be a large list of properties?  if that would be reasonable, then it might make more sense to use a Broadcast.

(separately, task serialization should probably avoid re-serializing the properties every time, but this could make that existing issue much worse,)

Reply:
Technically broadcast is faster than local properties if there are a lot of properties, but one problem is you need to carry the broadcast handler everywhere, which I don't think is applicable to `SQLConf.get`.

BTW we currently have hundreds of SQL configs, even a user set all of them for a job, the overhead is low. I tried
```
sc.makeRDD(Seq(1,2,3)).collect
1.to(100).foreach(i => sc.setLocalProperty(i.toString * 10, i.toString * 10))
sc.makeRDD(Seq(1,2,3)).collect
```
and didn't observe performance difference.

Reply:
we can avoid serializing local properties for each task, but that's a general optimization to local properties, we can do that in another PR.